core/src/test/java/com/linecorp/armeria/client/HttpResponseWrapperTest.java

```diff
@@ -0,0 +1,187 @@
+        final DecodedHttpResponse res = new DecodedHttpResponse(CommonPools.workerGroup().next());
+        final HttpResponseWrapper wrapper = httpResponseWrapper(res);
+
+        wrapper.tryWrite(HttpHeaders.of(200).addInt(HttpHeaderNames.CONTENT_LENGTH, "foo".length()));
+        wrapper.tryWrite(HttpData.ofUtf8("foo"));
+        wrapper.tryWrite(HttpHeaders.of(HttpHeaderNames.of("bar"), "baz"));
+        wrapper.tryWrite(HttpHeaders.of(HttpHeaderNames.of("qux"), "quux"));
+        wrapper.close();
+
+        final List<HttpObject> unaggregated = unaggregate(res);
+        assertThat(unaggregated).containsExactly(
+                HttpHeaders.of(200).addInt(HttpHeaderNames.CONTENT_LENGTH, 3),
+                HttpData.ofUtf8("foo"),
+                HttpHeaders.of(HttpHeaderNames.of("bar"), "baz"));
+    }
+
+    @Test
+    public void infromationalHeadersHeadersDataAndTrailers() throws Exception {
+        final DecodedHttpResponse res = new DecodedHttpResponse(CommonPools.workerGroup().next());
+        final HttpResponseWrapper wrapper = httpResponseWrapper(res);
+
+        wrapper.tryWrite(HttpHeaders.of(100));
+        wrapper.tryWrite(HttpHeaders.of(HttpHeaderNames.of("a"), "b"));
+        wrapper.tryWrite(HttpHeaders.of(200).addInt(HttpHeaderNames.CONTENT_LENGTH, "foo".length()));
+        wrapper.tryWrite(HttpData.ofUtf8("foo"));
+        wrapper.tryWrite(HttpHeaders.of(HttpHeaderNames.of("bar"), "baz"));
+        wrapper.close();
+
+        final List<HttpObject> unaggregated = unaggregate(res);
+        assertThat(unaggregated).containsExactly(
+                HttpHeaders.of(100),
+                HttpHeaders.of(HttpHeaderNames.of("a"), "b"),
+                HttpHeaders.of(200).addInt(HttpHeaderNames.CONTENT_LENGTH, "foo".length()),
+                HttpData.ofUtf8("foo"),
+                HttpHeaders.of(HttpHeaderNames.of("bar"), "baz"));
+    }
+
+    private static HttpResponseWrapper httpResponseWrapper(DecodedHttpResponse res) {
+        final HttpRequest req = HttpRequest.of(HttpMethod.GET, "/");
+        final ClientRequestContext cctx = ClientRequestContextBuilder.of(req).build();
+        final InboundTrafficController controller = InboundTrafficController.disabled();
+        final TestHttpResponseDecoder decoder =
+                new TestHttpResponseDecoder(cctx.log().channel(), controller);
+
+        res.init(controller);
+        return decoder.addResponse(1, req, res, cctx.logBuilder(), cctx.responseTimeoutMillis(),
+                                   cctx.maxResponseLength());
+    }
+
+    private static class TestHttpResponseDecoder extends HttpResponseDecoder {
+        TestHttpResponseDecoder(Channel channel,
+                                InboundTrafficController inboundTrafficController) {
+            super(channel, inboundTrafficController);
+        }
+    }
+
+    private static List<HttpObject> unaggregate(StreamMessage<HttpObject> req) throws Exception {
```

Comment:
What do you think about adding a new method to `StreamMessage`?

```java
public interface StreamMessage<T> ... {
    CompletableFuture<List<T>> drainAll();
    CompletableFuture<List<T>> drainAll(EventExecutor executor);
    CompletableFuture<List<T>> drainAll(boolean withPooledObjects);
    CompletableFuture<List<T>> drainAll(EventExecutor executor, boolean withPooledObjects);
}
```

Reply:
That's a good suggestion. Will add them. :)